code/trans.c

```diff
@@ -0,0 +1,377 @@
+                    transform,
+                    0, 1); /* these can't be old references */
+  if (res != ResOK)
+    goto failTable;
+
+  *transformReturn = transform;
+  return ResOK;
+
+failTable:
+  ControlFree(arena, transform, sizeof(TransformStruct));
+failAlloc:
+  return res;
+}
+
+
+void TransformDestroy(Transform transform)
+{
+  Arena arena;
+  Table oldToNew;
+
+  AVERT(Transform, transform);
+
+  /* TODO: Log some transform statistics. */
+
+  /* Workaround bootstrap problem, see .check.boot */
+  oldToNew = transform->oldToNew;
+  transform->oldToNew = NULL;
+  TableDestroy(oldToNew);
+
+  arena = TransformArena(transform);
+  transform->sig = SigInvalid;
+  ControlFree(arena, transform, sizeof(TransformStruct));
+}
+
+
+/* TransformArena -- return transform's arena
+ * 
+ * Must be thread-safe as it is called outside the arena lock. See
+ * <design/thread-safety/#sol.check>
+ */
+
+Arena TransformArena(Transform transform)
+{
+  Arena arena;
+  AVER(TESTT(Transform, transform));
+  arena = transform->arena;
+  AVER(TESTT(Arena, arena));
+  return arena;
+}
+
+
+Res TransformAddOldNew(Transform transform,
+                       Ref old_list[],
+                       Ref new_list[],
+                       Count count)
+{
+  Res res;
+  Index i;
+  Count added = 0;
+
+  AVERT(Transform, transform);
+  AVER(old_list != NULL);
+  AVER(new_list != NULL);
+  /* count: cannot check */
+  
+  res = TableGrow(transform->oldToNew, count);
+  if (res != ResOK)
+    return res;
+
+  for(i = 0; i < count; ++i) {
+    /* NOTE: If the mutator isn't adding references while the arena is parked,
+       we might need to access the client-provided lists, using ArenaRead. */
+    if(old_list[i] == NULL)
+      continue;  /* permitted, but no transform to do */
+    if(old_list[i] == new_list[i])
+      continue;  /* ignore identity-transforms */
+
+    /* Old refs must be in managed memory. */
+    {
+      Seg seg;
+      AVER(SegOfAddr(&seg, transform->arena, old_list[i]));
+    }
+
+    res = TableDefine(transform->oldToNew, (Word)old_list[i], new_list[i]);
+    AVER(res != ResFAIL); /* It's a static error to add the same old twice. */
+    if (res != ResOK)
+      return res;
+    
+    ++added;
+  }
+
+  AVERT(Transform, transform);
+  
+  return ResOK;
+}
+
+
+/* TransformApply -- transform references on the heap */
+
+static Res transformFix(Seg seg, ScanState ss, Ref *refIO)
+{
+  Ref ref;
+  Transform transform;
+  Res res;
+
+  AVERT_CRITICAL(Seg, seg);
+  AVERT_CRITICAL(ScanState, ss);
+  AVER_CRITICAL(refIO != NULL);
+
+  transform = ss->fixClosure;
+  AVERT_CRITICAL(Transform, transform);
+
+  if (!transform->aborted) {
+    void *refNew;
+
+    ref = *refIO;
+
+    if (TableLookup(&refNew, transform->oldToNew, (Word)ref)) {
+      if (ss->rank == RankAMBIG) {
+        /* We rely on the fact that ambiguous references are fixed
+           first, so that no exact references have been transformed
+           yet. */
```

Comment:
m. This dependency should be documented in the design and tagged, and referenced in the documentation of reference ranks. rule.code.justified, rule.code.deps, rule.code.design

Reply:
Fixed: Cross-referenced in de146ef65